flux-version: show broker, library, and command version, and add API version access

garlick opened the pull request:
This adds a new broker attribute `version` which reports the broker's compiled in version, e.g.
```
$ flux getattr version
0.10.0-469-g0605e1a0
```

Then it adds a  `--remote` option to `flux version` so that, in addition to reporting _its_ compiled in version, it can fetch the broker's version if desired:
```
$ flux version --remote
flux-core-0.10.0-469-g0605e1a0
```

We mused in #1810 that this might come in handy as our flux deployments get more complicated, and it was super simple to add, so here it is.

grondo commented:
Cool!

I think the `--remote` option could be confusing (especially if you are connected via a URI explicitly called `local://`), though I don't have a better idea (maybe --parent?)

My thought was that users may not know to run two separate commands in case the enclosing instance is a different version than the local `flux` command. What if `flux version` just looks for `FLUX_URI` set in environment and queries the "remote" version automatically, and reports both versions like:
```
commands: flux-core-x.y.z
broker: flux-core-x.y.z
```
It might also be useful for `flux version` to eventually report the version of `flux-security` to which it is linked, which might lend itself nicely to this extensible format or something similar.

We also have the open request for a `--verbose` flag in #1632. I'm not sure the best way to incorporate that into an expanded `flux version` output format.

grondo commented:
> That way there's always just a version on stdout, which might have scripting advantages over a more complicated output.

Yeah, I was thinking about machine readable output as well, but then I realized that maybe there is a better way to get the various versions if you need it in a script, and that the basic `flux version` output should be as helpful as possible for users, e.g.
```
Version of flux commands:
flux-core-x.y.z
Version of flux instance at URI=local:///tmp/flux-ftv4mnM:
flux-core-x.y.z +security
```

For machine readable output you could add a separate command or option. Since scriptable output would be run via scripts, it might make sense to make that invocation require the extra option, instead of the other way around.

garlick commented:
Those are good points!

Uh, we are getting the command version from a preprocessor symbol set in config.h.  What if we are dynamically linking against a different flux library?   Do we need a way to fetch that version as well?

garlick commented:
Maybe the ideal output would be:
```
libflux-core: version
libflux-security: version
commands: version
broker: vesion

chu11 commented:
> It might also be useful for flux version to eventually report the version of flux-security to which it is linked, which might lend itself nicely to this extensible format or something similar.

Looking even farther out, versions of loaded modules in the broker?  i.e. what version of flux-sched is loaded?

garlick commented:
> Looking even farther out, versions of loaded modules in the broker? i.e. what version of flux-sched is loaded?

Excellent point.  Maybe that should be an addition to `flux module list` with some well known (RFC) interface for obtaining the version of a loaded module?

grondo commented:
> Excellent point. Maybe that should be an addition to flux module list with some well known (RFC) interface for obtaining the version of a loaded module?

Yes, I like that idea, and module list protocol needs a rework anyway.

garlick commented:
OK, I added the macros and functions we discussed, and changed `flux-version` output so it looks like this
```
$ flux version
commands:    	0.10.0-506-g34e9b29a
libflux-core:	0.10.0-506-g34e9b29a
broker:  	0.10.0-506-g34e9b29a
FLUX_URI:	local:///tmp/flux-o3OqOQ
```
If FLUX_URI isn't set then broker and FLUX_URI are omitted from the output.

grondo commented on src/cmd/flux-start.c:
```diff
@@ -336,7 +336,7 @@ static void state_cb (flux_subprocess_t *p, flux_subprocess_state_t state)
                 log_msg ("%d (pid %d) %s", cli->rank, pid, strsignal (WTERMSIG (status)));
-            else if (WIFEXITED (status))
+            else if (WIFEXITED (status) && WEXITSTATUS (status) != 0)
                 log_msg ("%d (pid %d) exited with rc=%d", cli->rank, pid, WEXITSTATUS (status));
```
Hm, I had thought you'd already fixed this. Maybe in an existing PR that hasn't been merged?

While you are in here, the tests for `WIFSTOPPED()` (actually erroneously written here as `WIFSIGNALED()`) and `WIFCONTINUED()` should be removed. This code is only executed under `FLUX_SUBPROCESS_EXITED` so those states will never be present.

garlick replied:
Oh, hmm, I thought I had fixed it the way you asked before too....  I cherry picked that from another PR thinking it would get in quicker if I tacked it on here.  I'll revist, thanks for catching.

grondo commented on src/cmd/builtin/version.c:
```diff
@@ -26,9 +26,29 @@
+    flux_t *h = builtin_get_flux_handle (p);
+    if (!h)
+        log_err_exit ("flux_open");
+    if (!(version = flux_attr_get (h, "version", NULL)))
```
Rather than exiting with error on an invalid `FLUX_URI`, would it be better to either
 - silently ignore error from `flux_open` and act as if `FLUX_URI` wasn't set
 - or report a more user-friendly error `Failed to connect to broker at FLUX_URI=%s`
?

garlick replied:
I think at minimum a user friendly error should be reported since other things aren't going to do well if you have FLUX_URI set to a dead instance or whatever.

grondo replied:
Thanks, `flux start -s` will still work, but going forward that will not be a very common case, so I agree with your assesment...

garlick commented:
OK, I addressed those two comments and squashed.

grondo commented:
Excellent!

garlick commented:
Thanks!
